physical/zookeeper.go

```diff
@@ -194,3 +194,132 @@ func (c *ZookeeperBackend) List(prefix string) ([]string, error) {
 	sort.Strings(children)
 	return children, nil
 }
+
+// LockWith is used for mutual exclusion based on the given key.
+func (c *ZookeeperBackend) LockWith(key, value string) (Lock, error) {
+	l := &ZookeeperHALock{
+		in:    c,
+		key:   key,
+		value: value,
+	}
+	return l, nil
+}
+
+func (c *ZookeeperBackend) DetectHostAddr() (string, error) {
+	// TODO: implement this!
+	return "", nil
+}
+
+// ZookeeperHALock is a Zookeeper Lock implementation for the HABackend
+type ZookeeperHALock struct {
+	in    *ZookeeperBackend
+	key   string
+	value string
+
+	held     bool
+	leaderCh chan struct{}
+	zkLock *zk.Lock
+}
+
+func (i *ZookeeperHALock) Lock(stopCh <-chan struct{}) (<-chan struct{}, error) {
+	if i.held {
+		return nil, fmt.Errorf("lock already held")
+	}
+
+	// Attempt an async acquisition
+	didLock := make(chan struct{})
+	failLock := make(chan error, 1)
+	releaseCh := make(chan bool, 1)
+	lockpath := i.in.path + i.key
+	go func() {
+		// Wait to acquire the lock in ZK
+		acl := zk.WorldACL(zk.PermAll)
+		lock := zk.NewLock(i.in.client, lockpath, acl)
+		err := lock.Lock()
+		if err != nil {
+			failLock <- err
+			return
+		}
+		// Set node value
+		err2 := i.in.ensurePath(lockpath, []byte(i.value))
+		if err2 != nil {
+			failLock <- err2
+			lock.Unlock()
+			return
+		}
+		i.zkLock = lock
+
+		// Signal that lock is held
+		close(didLock)
+
+		// Handle an early abort
+		release := <-releaseCh
+		if release {
+			lock.Unlock()
+		}
+	}()
+
+	// Wait for lock acquisition, failure, or shutdown
+	select {
+	case <-didLock:
+		releaseCh <- false
+	case err := <-failLock:
+		return nil, err
+	case <-stopCh:
+		releaseCh <- true
+		return nil, nil
+	}
+
+	// Create the leader channel
+	i.held = true
+	i.leaderCh = make(chan struct{})
+
+	// Watch for Events which could result in loss of our zkLock and close(i.leaderCh)
+	currentVal, _, lockeventCh, err3 := i.in.client.GetW(lockpath)
+	if err3 != nil {
+		return nil, fmt.Errorf("unable to watch HA lock")
+	}
+	if i.value != string(currentVal) {
+		return nil, fmt.Errorf("lost HA lock immediately before watch")
+	}
+	go func() {
+		for {
+			select {
+			case event := <- lockeventCh:
+				// Lost connection?
+				if event.State != zk.StateConnected && event.State != zk.StateHasSession {
+					close(i.leaderCh)
+				}
+				// Lost watch
+				if event.Type == zk.EventNotWatching {
+					close(i.leaderCh)
+				}
+				// Lock changed
+				if event.Type == zk.EventNodeCreated || event.Type == zk.EventNodeDeleted || event.Type == zk.EventNodeDataChanged {
+					close(i.leaderCh)
+				}
+			}
+		}
+	}()
+
+	return i.leaderCh, nil
+}
+
+func (i *ZookeeperHALock) Unlock() error {
+	if !i.held {
+		return nil
+	}
+
+	close(i.leaderCh)
+	i.leaderCh = nil
+	i.held = false
+	i.zkLock.Unlock()
+	return nil
+}
+
+func (i *ZookeeperHALock) Value() (bool, string, error) {
+	lockpath := i.in.path + i.key
+	value, _, err := i.in.client.Get(lockpath)
+	return i.held, string(value), err
```

Comment:
The first bool should indicate if the lock is held by any node, not just this node. e.g. `true` if anybody holds the lock